modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -340,10 +351,19 @@ public void start() {
      */
     public void updateAllLocalStatistics() {
         try {
+            GridCompoundFuture<Boolean, Boolean> compoundFuture = new GridCompoundFuture<>(CU.boolReducer());
+
             distrMetaStorage.iterate(STAT_OBJ_PREFIX, (k, v) -> {
                 StatisticsObjectConfiguration cfg = (StatisticsObjectConfiguration)v;
 
-                updateLocalStatistics(cfg);
+                compoundFuture.add(updateLocalStatisticsAsync(cfg));
+            });
+
+            compoundFuture.markInitialized();
+
+            compoundFuture.listen(future -> {
```

Comment:
previous approach was blocking and here we not waiting execution is it ok here ?

Reply:
Blocking operation under metastorage read-lock was the reason of hanging,
because disco-notify thread requires metastorage write-lock to finish CAS operation future.